batch2/batch/database.py

```diff
@@ -9,3 +20,56 @@ def __init__(self, rv):
     if rv['rc'] != 0:
         raise CallError(rv)
     return rv
+
+
+class LeasedDatabase:
+    def __init__(self):
+        self.db = Database()
+        self.lease = asyncio.Event()
+        self.token = secrets.token_urlsafe(16)
+
+    async def acquire_lease(self):
+        while True:
+            try:
+                self.lease.clear()
+                now = time_msecs()
+                await check_call_procedure(
+                    self.db,
+                    'CALL acquire_lease(%s, %s, %s);',
+                    (self.token, now, now + 30 * 1000))
+                self.lease.set()
+            except asyncio.CancelledError:  # pylint: disable=try-except-raise
+                raise
+            except CallError:
+                log.exception('in acquiring lease')
+            finally:
+                await asyncio.sleep(15)
+
+    async def async_init(self, autocommit=True, maxsize=10):
+        await self.db.async_init(autocommit=autocommit, maxsize=maxsize)
+        self.lease.clear()
+        asyncio.ensure_future(self.acquire_lease())
+
+    async def just_execute(self, sql, args=None):
+        await self.lease.wait()
+        await self.db.just_execute(sql, args)
```

Comment:
OK, this is nice, but it doesn't _quite_ work.  Since there's a network between you and the database, you don't know how long it has been since the acquire actually succeeded.  If the response on the acquire message is delayed, the lease might have already expired.  We're also async, so there might have been other work scheduled between the lease set and the call to just_execute.

Few other ideas:

But!  I think a small modification of this will work.  You just need to include a lease check in a leased transaction you execute.  I think you can do this with a small modification to my PR adding transactions to Database: https://github.com/hail-is/hail/pull/7641.

The problem here is that Kubernetes is running two copies on update.  What we really want is at most one copy running at any time.  This is how a single-replicate stateful set behaves.  Kubernetes has to take down a pod to make the volume available to the upgraded pod.  If there is only one copy, you can't do no-downtime upgrades.  So we might take advantage of stateful sets and use the volume as the lease, so to speak.

If you do go the database lease solution, I think there a couple more issues you'll want to sort:  If you have the lease, loose it, someone else modifies the instances, and then you regain the lease, you're in the same boat.  One option is to obtain the lease when you start, and if you ever fail to keep it, you terminate yourself.

I'm not sure hanging is the right solution if you don't have the lease.  From the perspective of, say, a worker calling job complete, you probably want a temporary failure so the worker retries and gets the actual leader rather than just hanging.

Another option is to add some kind of heal function to the driver that reconciles divergence between memory and the database: refresh instances on errors, if we receive messages from unknown instances, we check if they are available in the database, etc.

Reply:
> This is how a single-replicate stateful set behaves. Kubernetes has to take down a pod to make the volume available to the upgraded pod.

This seems like the simplest option especially considering the other priorities we have. What happens if the upgrade fails? Does it try and spin up the old version again?

Reply:
I think the simplest, yes.  It will back off and retry like deployments, I assume.  We ultimately need rollback and host of other deployment issues, but that should be orthogonal to deployment vs stateful set.

Reply:
What about the recreate option on deployment?
https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#strategy

Reply:
I made a PR #7665